### ISSUE TITLE
Produces files should be checked when rebuilding

### ISSUE THREAD

**Issue by ndmitchell** [ndmitchell/shake#631]
At the moment if a rule `produces [foo.txt]` then next build run if `foo.txt` has been deleted it won't rebuild. It probably should. See https://github.com/ndmitchell/shake/pull/630#issuecomment-443463377 by @snowleopard

**Comment by ndmitchell** [comment]
`getDirectoryFiles` should never be called on generated files, so you should use the IO variant. The above will fail if you build it once, then change `docs.tar.gz` to contain an extra file - then `getDirectoryFiles` will have been called twice in one run with differing values.

**Comment by snowleopard** [comment]
@ndmitchell Yes, got it! 

So, I guess the following would be a correct way of writing this rule:

```haskell
"docs/README.md" %> \readme -> do
    need ["docs.tar.gz"]
    cmd ["tar", "zxvf", "docs.tar.gz", "-C" "."]
    docs <- liftIO $ getDirectoryFilesIO "docs" "*"
    produces (docs \\ [readme])
```

**Comment by ndmitchell** [comment]
Yep, looks good! I suspect `produces docs` is fine (although given I haven't yet given it a real semantics I'm not sure - but it seems reasonable). Also `cmd "tar zxvf docs.tar.gz -C ."` is equivalent to the `cmd` variant you wrote but more readable, so I prefer that.

**Comment by dwijnand** [comment]
This is a minimal reproduction:

```haskell
"bar.txt" %> \out -> do
    writeFile' out "bar"
    writeFile' "foo.txt" "foo"
    produces ["foo.txt"]
```

Any tips as to how to go about fixing this?  I don't know enough about core Shake to know how to fix this.  I see `produces` is implemented in terms of `trackWrite`, so it's not that...
